### DISCUSSION TITLE
How to validate an Object with optional properties ?

### DISCUSSION THREAD

**Discussion by EminDevNoth** [open-circle/valibot#713] [Q&A] [answered]
### What do we have ?
Let's take this simple example

```ts
type Data = {
    email?: string | undefined;
    password?: string | undefined;
}
```

### What is expected ?
I want to validate my data to be sure it is an **object**, and that each defined property is a string.

### What was done ?
```ts
const Schema = v.partial(
  v.object({
    email: v.string(),
    password: v.string(),
  }),
);
```

### The Issue
When I pass an array as data the test does not fail.

```ts
const result = v.safeParse(Schema, []);

// RESULT
[log]: {
  typed: true,
  success: true,
  output: {},
  issues: undefined
}
```

I don't know if I'm missing something or doing something wrong, but if anyone can enlighten me on this that would be great

**To test**: [Playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgGUmALAUxDI4AXkwA6MMljAyACnpxxECgCs+TGLIDeCxXAHJgpAFziOUYKwDmsgJQAaXYsmpUAd2gATUxjHnLNg66AL5BtgzMbBxwUHyoAK6k8KK+qMjYfAAKUqh8stz8gvZwANoAuuGMLOwQpHxipBA2sQlJlUA)

**Comment by fabian-hiller** [comment]
Any complex data structure extends from an object in JavaScript. This includes arrays. An array has all the properties of an object. Previously we had a strict check for plain objects, but this caused many problems in different environments. That's why we decided to only check for `typeof input === "object"`.

**Comment by EminDevNoth** [comment]
Okay, I see, it's a pretty restrictive situation I must say. 

Even though I don't like this solution, I will use this workaround for now
```ts
const Schema = v.pipe(
  v.custom((input) => input.constructor.name === "Object", "Should be an object"),
  v.partial(
    v.object({
      email: v.string(),
      password: v.string(),
    }),
  ),
);
```
> [!IMPORTANT]  
> For anyone who will be using this workaround, you should be aware that the order of the schema is important, as it appears that it is not the original object that is passed as input via the schemas (to verify, because I am not very familiar with this library), therefore it will not work if the schema `custom ` is not the first to be called

@fabian-hiller  Thank you for your reply. your responsiveness is crazy. I appreciate 🙏

**Reply by fabian-hiller** [reply, marked as answer]
I recommend using `input && typeof input === 'object' && input.constructor === Object`, since the name of a constructor can be easily manipulated. Here are related issues:  #587, #602, #608.
